azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ProxyConnectionHandler.java

```diff
@@ -0,0 +1,194 @@
+        }
+
+        final ErrorCondition errorCondition = transport.getCondition();
+        final String hostName = event.getReactor().getConnectionAddress(connection);
+        final ProxySelector proxySelector = ProxySelector.getDefault();
+
+        if (errorCondition == null
+                || !(errorCondition.getCondition().equals(ConnectionError.FRAMING_ERROR)
+                        || errorCondition.getCondition().equals(AmqpErrorCode.PROTON_IO_ERROR))
+                || proxySelector == null
+                || StringUtil.isNullOrEmpty(hostName)) {
+            return;
+        }
+
+        final String[] hostNameParts = hostName.split(":");
+        if (hostNameParts.length != 2) {
+            return;
+        }
+
+        int port;
+        try {
+            port = Integer.parseInt(hostNameParts[1]);
+        } catch (NumberFormatException ignore){
+            return;
+        }
+
+        final IOException ioException = reconstructIOException(errorCondition);
+        proxySelector.connectFailed(
+                createURIFromHostNamePort(this.getMessagingFactory().getHostName(), this.getProtocolPort()),
+                new InetSocketAddress(hostNameParts[0], port),
+                ioException);
+    }
+
+    @Override
+    public String getOutboundSocketHostName() {
+        final InetSocketAddress socketAddress = getProxyAddress();
+        return socketAddress.getHostString();
+    }
+
+    @Override
+    public int getOutboundSocketPort() {
+        final InetSocketAddress socketAddress = getProxyAddress();
+        return socketAddress.getPort();
+    }
+
+    private Map<String, String> getAuthorizationHeader() {
+        final PasswordAuthentication authentication = Authenticator.requestPasswordAuthentication(
+                getOutboundSocketHostName(),
+                null,
+                getOutboundSocketPort(),
+                null,
+                null,
+                "http",
+                null,
+                Authenticator.RequestorType.PROXY);
+        if (authentication == null) {
+            return null;
+        }
+
+        final String proxyUserName = authentication.getUserName();
+        final String proxyPassword = authentication.getPassword() != null
+                ? new String(authentication.getPassword())
+                : null;
+        if (StringUtil.isNullOrEmpty(proxyUserName)
+                || StringUtil.isNullOrEmpty(proxyPassword)) {
+            return null;
+        }
+
+        final HashMap<String, String> proxyAuthorizationHeader = new HashMap<>();
+        // https://tools.ietf.org/html/rfc7617
+        final String usernamePasswordPair = proxyUserName + ":" + proxyPassword;
+        proxyAuthorizationHeader.put(
+                "Proxy-Authorization",
+                "Basic " + Base64.getEncoder().encodeToString(usernamePasswordPair.getBytes()));
+        return proxyAuthorizationHeader;
+    }
+
+    private InetSocketAddress getProxyAddress() {
```

Comment:
>getProxyAddress [](start = 30, length = 15)

this method is called once to get host and again to get port. Call it once and use the same value.

Reply:
chose this way to simplify invoke order of `getOutboundSocketHostName()` and `getOutboundSocketHostPort()` methods.
If I fetch this value and cache - I would be assuming - one came before the other.
And - I did not want to cache the proxy properties - java allows for dynamic proxy configuration update - so always reading latest...

one other choice is to return hostname proxy pair. I preferred this over that - `readability` over  `efficiency` as the `getProxyAddress()` call is very fast and is not in hotpath.